examples/fungible-token/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_is(std::iter::once(&*INIT_ACTOR_ADDR))?;
+
+        let ConstructorParams {
+            ipc_gateway,
+            ipc_address,
+            name,
+            symbol,
+            balances,
+        } = params;
+
+        let balances = balances.into_iter().try_fold(HashMap::new(), |mut m, (a, b)| -> Result<_, ActorError> {
+            let id = rt
+                .resolve_address(&a)
+                .ok_or(actor_error!(illegal_argument; "cannot resolve address in initial balance table"))?
+                .id()
+                .unwrap();
+            m.insert(id, b);
+            Ok(m)
+        })?;
+
+        let st = State::new(rt.store(), ipc_gateway, ipc_address, name, symbol, balances).map_err(
+            |e| e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to create actor state"),
+        )?;
+        rt.create(&st)?;
+
+        Ok(())
+    }
+
+    fn name<BS, RT>(rt: &mut RT) -> Result<String, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        Ok(st.name().to_string())
+    }
+
+    fn symbol<BS, RT>(rt: &mut RT) -> Result<String, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        Ok(st.symbol().to_string())
+    }
+
+    fn total_supply<BS, RT>(rt: &mut RT) -> Result<TokenAmount, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        Ok(st.total_supply().clone())
+    }
+
+    fn balance<BS, RT>(rt: &mut RT, addr: Address) -> Result<TokenAmount, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let id = rt
+            .resolve_address(&addr)
+            .ok_or_else(|| actor_error!(illegal_argument; "cannot resolve account address"))?
+            .id()
+            .unwrap();
+        let st: State = rt.state()?;
+        let b = st.balance(rt.store(), id).map_err(|e| {
+            e.downcast_default(
+                ExitCode::USR_ILLEGAL_STATE,
+                "failed to get balance from store",
+            )
+        })?;
+        Ok(b)
+    }
+
+    fn transfer<BS, RT>(rt: &mut RT, params: TransferParams) -> Result<TransferReturn, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let TransferParams { to, amount } = params;
+
+        let from_id = rt.message().caller().id().unwrap();
+        let to_id = rt
+            .resolve_address(&to)
+            .ok_or_else(
+                || actor_error!(illegal_argument; "cannot resolve destination account address"),
+            )?
+            .id()
+            .unwrap();
+
+        let (from_balance, to_balance) = rt.transaction(|st: &mut State, rt| {
+            st.transfer(rt.store(), from_id, to_id, amount)
```

Comment:
Where do you check if the balance for the address is being used for an atomic execution. If I got the code correctly you can still use the balance locked for an execution for an internal transfer (not involved in an atomic exec), or did I missed something?

Reply:
[This](https://github.com/consensus-shipyard/ipc-atomic-execution/pull/6/files/e5d70adfecfd7883382fdb84ece2308027c5617f#diff-9496377190083568b9e1883497f39c8481a75733b86a346f1bbe8d08744cb824R270-R277) will fail if the sender balance is locked for an atomic transfer because `AtomicInputState` doesn't allow modifying the inner state while locked.

Reply:
Oh! So the check is implicit in `modify` (I missed that). Got it.